### ISSUE TITLE
haskell_cabal_library linker errors when using GHC 8.10

### ISSUE THREAD

**Issue by mboes** [tweag/rules_haskell#1418]
**Describe the bug**
Building even a simple `haskell_cabal_library` fails with linker errors when using GHC 8.10. The same code built without Cabal does not lead to linker errors.

**To Reproduce**
Switch `rules_haskell` to GHC 8.10.1 or 8.10.2. Then,

```
$ bazel build @stackage//:void
[undefined reference to various RTS symbols errors]
```

**Expected behavior**
I expect the build to succeed with no linker errors.

**Environment**
 - OS name + version: NixOS 20.03 (using a Nixpkgs toolchain) or Debian buster (using GHC bindists).

**Additional context**
This bug has already been reported multiple times: see #1401 #1413 #1327. I'm creating a new issue to gather what we know in a single place, with a more focused issue description.

According to https://github.com/tweag/rules_haskell/issues/1327#issuecomment-664378765, this is a regression introduced by #1388, which is itself necessary on some systems when using GHC 8.8. Since we don't use GHC 8.10 during CI testing, we didn't catch this.

On a PIE-by-default systems like Debian Buster,
* building a `haskell_library` using GHC 8.8 works fine, with or without #1388,
* building a `haskell_cabal_library` using GHC 8.8 works fine, with or without #1388,
* building a `haskell_library` using GHC 8.10 works fine, with or without #1388,
* building a `haskell_cabal_library` using GHC 8.10 works but only without #1388.

The latter case also fails on NixOS with a Nixpkgs toolchain.

So what's still unclear is why the result matrix for `haskell_cabal_library` is different on GHC 8.8 and on GHC 8.10. It's likely that https://gitlab.haskell.org/ghc/ghc/-/issues/15319 has something to do with it.

What's also unclear is why for GHC 8.10, the result matrix is different for `haskell_library` and `haskell_cabal_library`. The former works after #1388 while the latter does not.

cc @patrickt @robinbb

**Comment by robinbb** [comment]
I noticed during my investigation of this issue (on top of `github/semantic` repo) that the `ghc-cabal` binary shipped with the GHC bindist was not linked with the GHC-supplied 'libffi'. I wonder if this is related; perhaps the bindists are improperly built?

**Comment by mboes** [comment]
Additional findings:
* when building a library, with GHC 8.10 one must take care to *not* pass `-no-pie`. The reason is that this seems to confuse the linker into thinking that we want to produce a binary after all, so a shared library (although with PIE, the distinction between the two becomes pretty blurry).
* when building a binary, we now must make sure to pass `-no-pie` always. This is due to https://gitlab.haskell.org/ghc/ghc/-/issues/15319, as mentioned in the description.

I have a tentative hacky fix that now makes the entire rules_haskell test suite pass under GHC 8.10.2. Will clean up and submit tomorrow. What it does is change `haskell/cabal.bzl` to pass `-optl-no-pie` to GHC more selectively (only when building a binary). It also deletes the addition of `-no-pie` to the command-line when building libraries in `haskell/private/actions/compile.bzl`.

We really need a better heuristic in GHC than https://gitlab.haskell.org/ghc/ghc/-/issues/15319. Maybe there is still time to implement this ahead of GHC 9.0. As described in the description of that ticket, the ideal solution would be to be able to pass a `settings` file on the command-line. cc @monoidal

**Comment by mboes** [comment]
Oh and FTR: linker errors are the worst. ;)

This was a hard fought battle.

**Comment by mboes** [comment]
It's possible that #1424 will no longer be necessary for 8.10.3 and 8.8.5 according to https://gitlab.haskell.org/ghc/ghc/-/issues/17919, if those do get released. I'd rather tell users to upgrade to the next point release than to keep the gnarly logic introduced in #1424.

**Comment by monoidal** [comment]
The next versions of GHC (8.8.5, 8.10, 9.0.1) will contain a flag `-pgmc-supports-no-pie` that will permit simplifying the logic, https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4136#note_302255.
